proofs/refine.mli

```diff
@@ -19,15 +19,17 @@ open Proofview
 
 (** {7 Refinement primitives} *)
 
-val refine : typecheck:bool -> (Evd.evar_map -> Evd.evar_map * EConstr.t) -> unit tactic
-(** In [refine ~typecheck t], [t] is a term with holes under some
+val refine : ?with_classes:bool -> typecheck:bool -> (Evd.evar_map -> Evd.evar_map * EConstr.t) -> unit tactic
+(** In [refine ?(with_classes=true) ~typecheck t], [t] is a term with holes under some
     [evar_map] context. The term [t] is used as a partial solution
     for the current goal (refine is a goal-dependent tactic), the
     new holes created by [t] become the new subgoals. Exceptions
     raised during the interpretation of [t] are caught and result in
-    tactic failures. If [typecheck] is [true] [t] is type-checked beforehand. *)
+    tactic failures. If [typecheck] is [true] [t] is type-checked beforehand.
+    If [with_classes] is false then shelved evars are marked as unresolvable
+    in addition to goals. *)
```

Comment:
I read "in addition to goals", but could find only the code which marks the shelved evars as unresolvable. Where are goals marked as unresolvable?

Reply:
That's the `mark_as_goals` line in the primitive refine which performs this,  we always remove evars marked as goals from the set of typeclass evars.

Reply:
Ah! I see, thanks. And what does marking an evar as a goal do else? Couldn't we mark shelved goals as goals?

Reply:
(said otherwise, I'm wondering why we need to apply two different operations on generated goals, depending on whether they are shelved or not).

Reply:
That's a good point. The problem is that the code marking goals should not mark shelved evars as unresolvable always, precisely for the call to resolution above. We can certainly simplify that logic in the next step.

Reply:
Yeah, sure, I'd expect one operation being applied to two possible things (unshelved goals or all goals), depending on the flag. But here we have two operations being applied to two different things.

On a separate note, does `unshelve` mark evars as goals or unresolvable? Otherwise, I'd expect strange behaviors, like `refine; unshelve` not behaving the same as `simple refine` w.r.t. typeclasses.

Reply:
Maybe that possibility is not necessary and we should rather always "scope" the typeclass resolution to a single call to refine

Reply:
marking as a goal is the same as marking unresolvable.

Reply:
There's no `unshelve` tactic, but you're right that `unshelve refine foo`, in case there remains typeclass goals shelved would put them back in the goal list without removing them from the typeclass evar set.

Reply:
So it would seem that the same "shelving also implies not resolvable" semantics could be used for all variants of the refine tactic.

Reply:
Ah yes, it would be simpler, indeed.